src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs

```diff
@@ -103,6 +112,85 @@ public override async Task SendConnectionAsync(string connectionId, string metho
+            var task = _clientInvocationManager.Caller.AddInvocation<T>(connectionId, invocationId, cancellationToken);
+
+            // Exception handling follows https://source.dot.net/#Microsoft.AspNetCore.SignalR.Core/DefaultHubLifetimeManager.cs,349
+            try
+            {
+                return await task;
+            }
+            catch
+            {
+                _clientInvocationManager.Caller.RemoveInvocation(invocationId);
+                throw;
+            }
+        }
+
+        public override async Task SetConnectionResultAsync(string connectionId, CompletionMessage result)
+        {
+            if (IsInvalidArgument(connectionId))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+            }
+            if (_clientConnectionManager.ClientConnections.TryGetValue(connectionId, out var clientConnectionContext))
+            {
+                // Determine which manager (Caller / Router) the `result` belongs to.
+                // `TryCompletionResult` returns false when the corresponding invocation is not existing.
+                IClientResultsManager clientResultsManager = null;
+                if (_clientInvocationManager.Caller.TryCompleteResult(connectionId, result))
+                {
+                    clientResultsManager = _clientInvocationManager.Caller;
+                }
+                if (_clientInvocationManager.Router.TryCompleteResult(connectionId, result))
+                {
+                    clientResultsManager = _clientInvocationManager.Router;
+                }
+
+                // Block unknown `results` which belongs to neither Caller nor Router
+                if (clientResultsManager != null)
+                {
+                    var protocol = clientConnectionContext.Protocol;
+                    // For router server, it should send a ClientCompletionMessage with accurate payload content, which is necessary for the caller server.
+                    // For caller server, the only purpose of sending ClientCompletionMessage is to inform service to cleanup the invocation, which means only InvocationId and ConnectionId make sense. To avoid serialization for useless payload, we assign payload with empty bytes.
+                    var payload = clientResultsManager == _clientInvocationManager.Router
+                                ? SerializeCompletionMessage(result, protocol)
+                                : Array.Empty<byte>();
+
+                    var message = AppendMessageTracingId(new ClientCompletionMessage(result.InvocationId, connectionId, _callerId, protocol, payload));
+                    await WriteAsync(message);
+                }
+            }
+        }
+
+        public override bool TryGetReturnType(string invocationId, [NotNullWhen(true)] out Type type)
+        {
+            if (_clientInvocationManager.Router.ContainsInvocation(invocationId))
```

Comment:
Is this check necessary? Simply try get from caller then try get from router?

Reply:
Removed this check. Also remove method `Router.ContainsInvocation`